Updating shell.nix to flakes (#114)
godot-rust/gdnative-book: src/recipes/nix-build-system.md

# Recipe: Nix as development environment

> **Disclaimer**: _Currently the following steps are tested and confirmed to work on NixOs only._

Reviewer: ```suggestion
**Disclaimer**: _Currently the following steps are tested and confirmed to work on NixOS only._
```

Reviewer: Capitalization is also incorrect in some other places in this page, maybe search & replace 🙂

**Disclaimer**: _Currently the following steps are tested and confirmed to work on NixOs only._

> [Nix](https://nixos.org/) is a package manager that employs a pure functional approach to dependency management. Nix packages are built and ran in isolated environments. It makes them more portable, but also harder to author. This tutorial will walk you through the process of setting up a package for Godot, GDNative and Rust with Nix.

Reviewer: Why did you remove the introduction about NixOS? I don't think Flakes should be explained before NixOS itself.

Author: a little redundant since the relation is with flakes feature not nix package manager. But this doesnt change nothing after all

Reviewer 2: We shouldn't remove this.

> This tutorial assumes that you are on a [NixOs](https://nixos.wiki/wiki/Main_Page) system and have [Flakes](https://nixos.wiki/wiki/Flakes) enabled.

Reviewer 2: ```suggestion
This tutorial assumes that you are on a [NixOs](https://nixos.wiki/wiki/Main_Page) system
or that Nix is [installed](https://nixos.org/download.html#nix-quick-install).
Finally, you also need to make sure that [Flakes](https://nixos.wiki/wiki/Flakes) enabled.
```

          # Point bindgen to where the clang library would be
          LIBCLANG_PATH = "${pkgs.libclang.lib}/lib";
          # Make clang aware of a few headers (stdbool.h, wchar.h)
          BINDGEN_EXTRA_CLANG_ARGS = with pkgs; ''
            -isystem ${llvmPackages.libclang.lib}/lib/clang/${lib.getVersion clang}/include
            -isystem ${llvmPackages.libclang.out}/lib/clang/${lib.getVersion clang}/include
            -isystem ${glibc.dev}/include
          ''
          # Setup all the packages that we will need to develop with gdnative.
          buildInputs = [
            openssl
            pkg-config
            rust-bin.stable.latest.default
            godot3
          ];
        };
      }
    );
}
```

If you get any errors about missing headers, you can use [`nix-locate`](https://github.com/bennofs/nix-index#usage) to search for them, e.g. `nix-locate 'include/wchar.h' | grep -v '^('` (the `grep -v` hides indirect packages), and then add the matching Nix package via the `BINDGEN_EXTRA_CLANG_ARGS` env var like above ([context](https://github.com/NixOS/nixpkgs/issues/52447#issuecomment-853429315)).

## Activating the Flake environment

> In the same directory that you made your flake.nix, you can use the comand ['nix develop'](https://nixos.wiki/wiki/Development_environment_with_nix-shell#nix-develop) to load our flake config and setup everything for your NixOs system.

Reviewer: ```suggestion
In the same directory where you created your flake.nix, you can use the comamnd [`nix develop`](https://nixos.wiki/wiki/Development_environment_with_nix-shell#nix-develop) to load our Flake config and setup everything for your NixOS system.
```

> In the same directory where you created your flake.nix, you can use the comamnd [`nix develop`](https://nixos.wiki/wiki/Development_environment_with_nix-shell#nix-develop) to load our Flakes config and setup everything for your NixOS system.

Reviewer 2: ```suggestion
In the same directory where you created your flake.nix, you can use the command [`nix develop`](https://nixos.wiki/wiki/Development_environment_with_nix-shell#nix-develop) to load our Flakes config and setup everything for your Nix-capable system.
```

Nix is available in other distros as well

Author: Yes, but i didnt include that because of nixgl.
Nixgl is some graphics drivers workarounds for non-nixos nix package manager and it looks like they does not have a good flakes support.
